xonsh/proc.py

```diff
@@ -234,8 +216,256 @@ def __init__(self, fd, buffer=None, chunksize=1024):
+
+
+def _expand_console_buffer(cols, max_offset, expandsize, orig_posize, fd):
+    # if we are getting close to the end of the console buffer,
+    # expand it so that we can read from it successfully.
+    rows = ((max_offset + expandsize)//cols) + 1
+    winutils.set_console_screen_buffer_size(cols, rows, fd=fd)
+    orig_posize = orig_posize[:3] + (rows,)
+    max_offset = (rows - 1) * cols
+    return rows, max_offset, orig_posize
+
+
+def populate_console(reader, fd, buffer, chunksize, queue, expandsize=None):
+    """Reads bytes from the file descriptor and puts lines into the queue.
+    The reads happend in parallel, using xonsh.winutils.read_console_output_character(),
+    and is thus only available on windows. If the read fails for any reason, the reader is
+    flagged as closed.
+    """
+    # OK, so this function is super annoying because Windows stores its
+    # buffers as a 2D regular, dense array -- without trailing newlines.
+    # Meanwhile, we want to add *lines* to the queue. Also, as is typical
+    # with parallel reads, the entire buffer that you ask for may not be
+    # filled. Thus we have to deal with the full generality.
+    #   1. reads may end in the middle of a line
+    #   2. excess whitespace at the end of a line may not be real, unless
+    #   3. you haven't read to the end of the line yet!
+    # So there are alignment issues everywhere.  Also, Windows will automatically
+    # read past the current cursor position, even though there is presumably
+    # nothing to see there.
+    #
+    # These chunked reads basically need to happen like this because,
+    #   a. The default buffer size is HUGE for the console (90k lines x 120 cols)
+    #      as so we can't just read in everything at the end and see what we
+    #      care about without a noticible performance hit.
+    #   b. Even with this huge size, it is still possible to write more lines than
+    #      this, so we should scroll along with the console.
+    # Unfortnately, because we do not have control over the terminal emulator,
+    # It is not possible to compute how far back we should set the begining
+    # read position because we don't know how many characters have been popped
+    # off the top of the buffer. If we did somehow know this number we could do
+    # something like the following:
+    #
+    #    new_offset = (y*cols) + x
+    #    if new_offset == max_offset:
+    #        new_offset -= scolled_offset
+    #        x = new_offset%cols
+    #        y = new_offset//cols
+    #        continue
+    #
+    # So this method is imperfect and only works as long as the sceen has
+    # room to expand to.  Thus the trick here is to expand the screen size
+    # when we get close enough to the end of the screen. There remain some
+    # async issues related to not being able to set the cursor position.
+    # but they just affect the alignment / capture of the output of the
+    # first command run after a screen resize.
```

Comment:
Arghh.. Crazy. I knew windows behaved this way, but I must say I ca hardly follow the code below.